### PR TITLE
Rendering: Blinn-Phong

### PR DESCRIPTION
**Pull request by sp0lsh**

### CODE REVIEW

File: PolyEngine/Engine/Res/Shaders/blinn-phongFrag.shader [hunk 1]
```diff
@@ -1,9 +1,10 @@
 #version 330 core
 
-uniform sampler2D ourTexture;
+uniform sampler2D uTexture;
```
Comment (MuniuDev) [inline PolyEngine/Engine/Res/Shaders/blinn-phongFrag.shader]: Nice change ;)

File: PolyEngine/Engine/Res/Shaders/blinn-phongFrag.shader [hunk 2]
```diff
@@ -36,48 +44,73 @@ in vec3 vNormal;
...
 	DiffuseLight base = light.Base;
 	vec3 dir = light.Direction.xyz;
 	float dirNdotL = max(dot(normalWS, dir), 0.0);
-	vec4 dirDiffuse = vec4(base.Color.rgb, 1.0) * max(base.Intensity, 0.0) * dirNdotL;
-	return dirDiffuse;
+	result.Diffuse = vec3(base.Color.rgb) * max(base.Intensity, 0.0) * dirNdotL;
+	
+	vec3 reflectDir = reflect(-dir, normalWS);
+	float spec = pow(max(dot(uCameraDirection.xyz, reflectDir), 0.0), 32);
+	result.Specular = uSpecularStrength * spec * base.Color.rgb;
```
Comment (MuniuDev) [inline PolyEngine/Engine/Res/Shaders/blinn-phongFrag.shader]: Specular color is not always equal to base color. Specular color should be taken from the material. Correct me if i'm wrong.

File: PolyEngine/Engine/Src/Engine.cpp [hunk 3]
```diff
@@ -44,7 +44,7 @@ void Poly::Engine::Init(std::unique_ptr<IGame> game, std::unique_ptr<IRenderingD
 	RegisterWorldComponent<DeferredTaskWorldComponent>((size_t)eEngineWorldComponents::DEFERRED_TASK);
 	RegisterWorldComponent<SoundWorldComponent>((size_t) eEngineWorldComponents::SOUND);
 	RegisterWorldComponent<Physics2DWorldComponent>((size_t)eEngineWorldComponents::PHYSICS_2D);
-	RegisterWorldComponent<DiffuseLightSourceWorldComponent>((size_t)eEngineWorldComponents::DIFFUSE_LIGHTSOURCE);
+	RegisterWorldComponent<AmbientLightWorldComponent>((size_t)eEngineWorldComponents::DIFFUSE_LIGHTSOURCE);
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/Engine.cpp]: If we're changing class names, let's change enum field names as well.

File: PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp [hunk 4]
```diff
@@ -33,48 +37,56 @@ void RenderingSandbox::Init()
...
-	playerTransform->SetLocalTranslation(Vector(-3.0f, 0.0f, 5.0f));
-	playerTransform->SetLocalScale(10.0f);
-	playerTransform->SetLocalRotation(Quaternion(Vector::UNIT_Y, -90_deg) * Quaternion(Vector::UNIT_X, -90_deg));
+	// auto player = DeferredTaskSystem::SpawnEntityImmediate(world);
+	// DeferredTaskSystem::AddComponentImmediate<Poly::TransformComponent>(world, player);
+	// DeferredTaskSystem::AddComponentImmediate<Poly::MeshRenderingComponent>(world, player, "Models/tank2/bradle.3ds", eResourceSource::GAME);
+	// Poly::TransformComponent* playerTransform = world->GetComponent<Poly::TransformComponent>(player);
+	// playerTransform->SetLocalTranslation(Vector(-2.0f, 0.0f, 0.0f));
+	// playerTransform->SetLocalScale(10.0f);
+	// playerTransform->SetLocalRotation(Quaternion(Vector::UNIT_Y, -90_deg) * Quaternion(Vector::UNIT_X, -90_deg));
```
Comment (MuniuDev) [inline PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp]: Remove commented code. You can always bring it back using git.

File: PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp [hunk 5]
```diff
@@ -86,9 +98,20 @@ void RenderingSandbox::Init()
 	DeferredTaskSystem::AddComponentImmediate<Poly::MeshRenderingComponent>(world, ground, "Models/ground/ground.fbx", eResourceSource::GAME);
 	Poly::TransformComponent* groundTrans = world->GetComponent<Poly::TransformComponent>(ground);
 	//	groundTransform->SetLocalTranslation(Vector(SCALE * SIZE, 0.f, SCALE * SIZE));
-	groundTrans->SetLocalScale(SCALE);
+	// groundTrans->SetLocalScale(SCALE);
 
 	Engine->RegisterGameUpdatePhase(GameMainSystem::GameUpdate);
+
+	for (auto cmpTuple : world->IterateComponents<PostprocessSettingsComponent>())
```
Comment (MuniuDev) [inline PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp]: Looks hackish. Wouldn't it be better to just retrive the postprocess component from the object we're interested in?

File: PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp [hunk 6]
```diff
@@ -100,38 +123,117 @@ void RenderingSandbox::Deinit()
...
-	float Distortion	= 0.5f;
-	float Bloom			= 0.4f;
-	float ColorTempValue = 9500.0f; // 6500.0 from [1000.0, 40000.0]
-	float Saturation	= 0.5f;
-	float Grain			= 0.5f;
-	float Stripes		= 0.5f;
-	float Vignette		= 0.5f;
-	float Brightness	= 0.1f;
-	float Contrast		= 1.5f;
+void RenderingSandbox::AddDirectionalLights()
```
Comment (MuniuDev) [inline PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp]: This method is not used anymore, either remove it or move the implementation from `Init()` here.

File: PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp [hunk 7]
```diff
@@ -100,38 +123,117 @@ void RenderingSandbox::Deinit()
...
+	// 
+	// 	postCmp->Distortion = t * Distortion;
+	// 	postCmp->ColorTempValue = Lerp(6500.0f, ColorTempValue, t);
+	// 	postCmp->Saturation = Lerp(1.0f, Saturation, t);
+	// 	postCmp->Grain = Lerp(0.0f, Grain, t);
+	// 	postCmp->Stripes = Lerp(0.0f, Stripes, t);
+	// 	postCmp->Vignette = Lerp(0.0f, Vignette, t);
+	// 
+	// 	gConsole.LogInfo("PostCmp: t: {}, Saturation: {}, Vignette: {}", t, postCmp->Saturation, postCmp->Vignette);
+	// }
```
Comment (MuniuDev) [inline PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp]: Commented blocks of code should never be pushed to repository.

File: PolyEngine/RenderingDevice/OpenGL/Src/GLShaderProgram.hpp [hunk 8]
```diff
@@ -66,5 +66,9 @@ namespace Poly {
 		std::map<String, OutputInfo> Outputs;
 		GLuint ProgramHandle;
 		EnumArray<String, eShaderUnitType> ShaderCode;
+
+		String VertexProgramPath;
+		String GeometryProgramPath;
+		String FragmentProgramPath;
```
Comment (MuniuDev) [inline PolyEngine/RenderingDevice/OpenGL/Src/GLShaderProgram.hpp]: These three fields are never used. Remove them.

File: PolyEngine/Engine/Res/Shaders/blinn-phongFrag.shader (hunk 1 above)
Comment (sp0lsh) [reply PolyEngine/Engine/Res/Shaders/blinn-phongFrag.shader]: Thanks

File: PolyEngine/Engine/Src/Engine.cpp (hunk 3 above)
Comment (sp0lsh) [reply PolyEngine/Engine/Src/Engine.cpp]: done

File: PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp (hunk 4 above)
Comment (sp0lsh) [reply PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp]: Done, sorry for that

File: PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp (hunk 5 above)
Comment (sp0lsh) [reply PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp]: done, yes

File: PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp (hunk 6 above)
Comment (sp0lsh) [reply PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp]: done

File: PolyEngine/RenderingDevice/OpenGL/Src/GLShaderProgram.hpp (hunk 8 above)
Comment (sp0lsh) [reply PolyEngine/RenderingDevice/OpenGL/Src/GLShaderProgram.hpp]: done

File: PolyEngine/Engine/Res/Shaders/blinn-phongFrag.shader (hunk 2 above)
Comment (sp0lsh) [reply PolyEngine/Engine/Res/Shaders/blinn-phongFrag.shader]: Debugging it right now. I've spotted in that version wrongly added multiple lights influence resulting in grayish objects in bright scenes. Expect Blinn-Phong pt. 2 with fixes/improvements.

File: PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp (hunk 7 above)
Comment (sp0lsh) [reply PolyEngine/Games/SandboxRendering/Src/RenderingSandbox.cpp]: Sorry

**Pull request merged**